File: 11-纹理映射(下)/core/Mesh.java
package core;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;

public class Mesh {
	
	public Vector3D[] vertices;
	public Vector3D[] normals;
	public int[] indices;
	public float[][] uvCoordinates;
	public Vector3D tempVector = new Vector3D(0,0,0);
	
	public Mesh(String objFilePath, String order) {
		
		//构建文件路径
		String filePath = MainThread.class.getProtectionDomain().getCodeSource().getLocation().getPath();
		objFilePath = filePath + objFilePath;

		
		//从文件中读取顶点，法线的个数
        int numVertices = 0;
        int numNormals = 0;
        int numIndices = 0;
        int numUVs = 0;

        try (BufferedReader br = new BufferedReader(new FileReader(objFilePath))) {
            String line;
            while ((line = br.readLine()) != null) {
                if (line.startsWith("v ")) {
                    numVertices++;
                } else if (line.startsWith("vn ")) {
                    numNormals++;
                } else if (line.startsWith("f ")) {
                    numIndices += 3;
                } else if(line.startsWith("vt ")) {
                	numUVs++;
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        }
		
        //创建用来存储顶点位置， 法线的数组
        vertices = new Vector3D[numVertices];
        normals = new Vector3D[numVertices];
        indices = new int[numIndices];
        uvCoordinates = new float[numIndices][];
        
        //创建一个单独数组用来帮助调整法线数和UV数组的组索引， 这样顶点位置和法线和UV数组的索引就能对应上来
        Vector3D[] normalsTemp = new Vector3D[numNormals];
        float[][] uvCoordinatesTemp = new float[numUVs][];
        
        //再重新读一边这个 OBj文件
        try (BufferedReader br = new BufferedReader(new FileReader(objFilePath))) {
        	String line;
        	int vertexIndex = 0;
        	int normalIndex = 0;
        	int indexIndex = 0;
        	int uvIndex = 0;

        	while ((line = br.readLine()) != null) {
        		if (line.startsWith("v ")) {
        			String[] values = line.split("\\s+");        //读取顶点位置
        			float x = Float.parseFloat(values[1]);
        			float y = Float.parseFloat(values[2]);
        			float z = Float.parseFloat(values[3]);
        			vertices[vertexIndex++] = new Vector3D(x, y, z);
        		} else if (line.startsWith("vn ")) {
        			String[] values = line.split("\\s+");        //读取法线位置，存放在临时的法线数组中
        			float x = Float.parseFloat(values[1]);
        			float y = Float.parseFloat(values[2]);
        			float z = Float.parseFloat(values[3]);
        			normalsTemp[normalIndex++] = new Vector3D(x, y, z);
        		}else if(line.startsWith("vt ")) {
        			String[] values = line.split("\\s+"); 
        			float u = Float.parseFloat(values[1]);
        			float v = Float.parseFloat(values[2]);
        			uvCoordinatesTemp[uvIndex++] = new float[]{u,v};
        		}else if (line.startsWith("f ")) {
        			String[] values = line.split("\\s+");
        			
        			if(order != "clockwise") {
	        			for (int i = 3; i >= 1; i--) {
	        				String[] indicesValues = values[i].split("/");
	        				int vertexIndexValue = Integer.parseInt(indicesValues[0]) - 1;
	        				int uvIndexValue = Integer.parseInt(indicesValues[1]) - 1;
	        				int normalIndexValue = Integer.parseInt(indicesValues[2])-1;
	        				normals[vertexIndexValue] = normalsTemp[normalIndexValue];
	        				uvCoordinates[indexIndex] = uvCoordinatesTemp[uvIndexValue];
	        				indices[indexIndex++] = vertexIndexValue;
	        			}
        			}else {
        				 for (int i = 1; i <= 3; i++) {
  	                        String[] indicesValues = values[i].split("/");
  	                        int vertexIndexValue = Integer.parseInt(indicesValues[0]) - 1;
  	                        int uvIndexValue = Integer.parseInt(indicesValues[1]) - 1;
  	                        int normalIndexValue = Integer.parseInt(indicesValues[2])-1;
  	                        normals[vertexIndexValue] = normalsTemp[normalIndexValue];
  	                        uvCoordinates[indexIndex] = uvCoordinatesTemp[uvIndexValue];
  	                        indices[indexIndex++] = vertexIndexValue;
  	                    }
        			}
        		}
        	}
        } catch (IOException e) {
        	e.printStackTrace();
        }
        
        for(int i = 0; i < normals.length; i++) {
        	if(normals[i] == null)
        		normals[i] = new Vector3D(0,0,1);
        }
	}
	
	
	
}

File: SDL_demo/Mesh.cpp
#include <iostream>
#include <fstream>
#include <Windows.h>

#include "Mesh.h"

Mesh::Mesh(const std::string& objFilePath, const std::string& order) {
    // 构造文件路径
    char buffer[MAX_PATH];
    DWORD length = GetCurrentDirectoryA(MAX_PATH, buffer);

    if (length == 0) {
        std::cout << "获取当前目录失败。" << std::endl;
    }

    std::string currentPath(buffer);
    std::string filePath = currentPath + "/" + objFilePath;

    // 读取文件以计算顶点、法线、索引和 UV 的数量
    numVertices = 0;
    numNormals = 0;
    numIndices = 0;
    numUVs = 0;
        
    std::ifstream file(filePath);
    std::string line;
        
    while (std::getline(file, line)) {
        if (line.rfind("v ", 0) == 0) {
            numVertices++;
        } else if (line.rfind("vn ", 0) == 0) {
            numNormals++;
        } else if (line.rfind("f ", 0) == 0) {
            numIndices += 3;
        } else if (line.rfind("vt ", 0) == 0) {
            numUVs++;
        }
    }

    
    // 创建顶点、法线、索引和 UV 的数组
    vertices = new Vector3D[numVertices];
    normals = new Vector3D[numVertices];
    indices = new int[numIndices];
    uvCoordinates = new float*[numIndices];

        
    // 创建临时数组以辅助调整法线和 UV 索引
    Vector3D* normalsTemp = new Vector3D[numNormals];
    float** uvCoordinatesTemp = new float*[numUVs];
        
    
    // 再次读取 OBJ 文件以填充数组
    file.clear();
    file.seekg(0);

        
    int vertexIndex = 0;
    int normalIndex = 0;
    int indexIndex = 0;
    int uvIndex = 0;
        
    while (std::getline(file, line)) {
        if (line.rfind("v ", 0) == 0) {
            float x, y, z;
            sscanf(line.c_str(), "v %f %f %f", &x, &y, &z);
            vertices[vertexIndex++].set(x, y, z);
        } else if (line.rfind("vn ", 0) == 0) {
            float x, y, z;
            sscanf(line.c_str(), "vn %f %f %f", &x, &y, &z);
            normalsTemp[normalIndex++].set(x, y, z);
        } else if (line.rfind("vt ", 0) == 0) {
            float u, v;
            sscanf(line.c_str(), "vt %f %f", &u, &v);
            uvCoordinatesTemp[uvIndex++] = new float[2]{ u, v };
        } else if (line.rfind("f ", 0) == 0) {
            int indicesValues[9];
            sscanf(line.c_str(), "f %d/%d/%d %d/%d/%d %d/%d/%d",
                   &indicesValues[0], &indicesValues[1],
                   &indicesValues[2], &indicesValues[3],
                   &indicesValues[4], &indicesValues[5],
                   &indicesValues[6], &indicesValues[7],
                   &indicesValues[8]);

           
          
            if (order != "clockwise") {
                for (int i = 2; i >= 0; i--) {
                    int vertexIndexValue = indicesValues[i * 3] - 1;
                    int uvIndexValue = indicesValues[i * 3 + 1] - 1;
                    int normalIndexValue = indicesValues[i * 3 + 2] - 1;
                    normals[vertexIndexValue].set(normalsTemp[normalIndexValue]);
                    uvCoordinates[indexIndex] = uvCoordinatesTemp[uvIndexValue];
                    indices[indexIndex++] = vertexIndexValue;
                }
            } else {
                for (int i = 0; i < 3; i++) {
                    int vertexIndexValue = indicesValues[i * 3] - 1;
                    int uvIndexValue = indicesValues[i * 3 + 1] - 1;
                    int normalIndexValue = indicesValues[i * 3 + 2] - 1;
                    normals[vertexIndexValue].set(normalsTemp[normalIndexValue]);
                    uvCoordinates[indexIndex] = new float[2];
                    uvCoordinates[indexIndex][0] = uvCoordinatesTemp[uvIndexValue][0];
                    uvCoordinates[indexIndex][1] = uvCoordinatesTemp[uvIndexValue][1];
                    indices[indexIndex++] = vertexIndexValue;
                    
                }
            }
        }
    }

 

    // 清理临时数组
    delete[] normalsTemp;
    for (int i = 0; i < numUVs; i++) {
        delete[] uvCoordinatesTemp[i];
    }
    delete[] uvCoordinatesTemp;
    
        
    // 关闭文件
    file.close();

    
}

Mesh::~Mesh() {
    delete[] vertices;
    delete[] normals;
    delete[] indices;
    for (int i = 0; i < sizeof(uvCoordinates); i++) {
        delete[] uvCoordinates[i];
    }
    delete[] uvCoordinates;
}
